CONTRIBUTING.md

```diff
@@ -176,6 +176,156 @@ The Remote - Containers extension ships into **VS Code Insiders** as new feature
+- EG `_VSC_INSTALL_AZURE_CLI="azcli-debian.sh ${_BUILD_ARG_AZURE_CLI_VERSION:-latest}"`
+
+**Feature testing**
+
+*Local testing*
+
+- Create a devcontainer with a target base image.
+- Add your script to the root.
+- Bring up the container image.
+- Run your script in the container with required arguments.
+- Verify expected results.
+- Bring down container to clean up.
+
+Repeat as needed to iterate from a clean workspace.
+
+*Unit tests*
+- Add your feature to the [run-scripts.sh](script-library/test/regression/run-scripts.sh) file to ensure it is included in CI tests.
+- Your addition should take the form `runScript <feature> <non-default-args>`. EG `runScript dotnet "3.1 true ${USERNAME} false /opt/dotnet dotnet"`
+- If your script takes the installation user as an argument, be sure to specify it as ${USERNAME} in the tests for programatic testing.
+
+*Regression tests*
+- Add your feature to the [test-features.env](script-library/container-features/test-features.env) file to include it in regression tests of the container-feature functionality. By setting the `_VSC_INSTALL_<FEATURE>` ENV VAR to true and adding the expected _BUILD_ARG options for your feature.
+
+EG
+```
+    _VSC_INSTALL_DOTNET=true
+    _BUILD_ARG_DOTNET_VERSION=latest
+    _BUILD_ARG_DOTNET_RUNTIMEONLY=false
+```
+
+**Feature documentation**
+
+Add your new feature to the list of scripts in the [script-library README.md](https://github.com/microsoft/vscode-dev-containers/blob/main/script-library/README.md#scripts).
+
+Add documentation for your new feature script to the [script-library/docs](https://github.com/microsoft/vscode-dev-containers/blob/main/script-library/docs) directory.
+
+Documentation should include: 
+- the status of the script, supported operating systems, and maintainer.
+- the syntax expected to run as a feature or script
+- a description of the script arguments
+- detailed usage instructions
+
+Feel free to use other scripts in that directory as inspiration.
+
+### Best practices for writing feature install scripts
+
+- Decouple sections of the shellscript that handle user setup, helper functions, and feature installation. Doing so will apply a logical and natural flow to the script for future developers and maintainers to follow. One way to denote this distinction is to use in-line comments throughout the script.
+    ```md
+    # Logical flow recommended:
+    1. File header and description.
+    2. Define constants and default values.
+    3. User setup and user validation.
+    4. Helper functions.
+    5. Checks for dependencies being installed or installs dependencies.
+    6. Runs container feature installs.
+    7. Gives the user correct permissions if necessary.
+    ```
+
+- Error messages from the script should echo to `STDERR`, while regular status messages should echo to `STDOUT`. This makes troubleshooting the script easier for developers. One way to facilitate this is to create an `err()` function like so:
```

Comment:
Not sure if this whole section of how to "write" scripts makes sense.
This is a grey area where it depends of the shell type selected (sh, bash, ksh, zsh, ...) and the container that supports it, so some of these recommendations are specific to a shell type.

Reply:
Can you be more specific about which recommendations you have concerns about?

Reply:
@juzuluag for more context, we based this on the formatting and standards of the already existing shell scripts in the repo, and general best practices followed.  If you see some points that are too specific to a shell type, would you call those out in particular? We can refactor this section accordingly.

Reply:
Hi @bderusha @Kpopovs1, my comment is more towards to be flexible towards how to do things. The way I read this is like a 'forcing' howto do it. devcontainers have a lot of different containers types(i.e ubuntu, Debian, alpine), so the tooling might not always present. My suggestion is more towards examples of how to capture things (i.e stderr, catch errors) specifying that you were using xyz shell.

Reply:
I think I see juzuluag's point -- that even though 99% of the `script-library` scripts use bash, this new doc section seems not state as much up front, and that the guidance "may vary"

Reply:
I've refactored the parts of this section that talk about specific code best practices to emphasize that they are recommendations if using bash, let me know if it's clearer now @juzuluag and @dnastrain